clients/roscpp/src/libros/spinner.cpp

```diff
@@ -30,10 +30,90 @@
 #include "ros/callback_queue.h"
 
 #include <boost/thread/thread.hpp>
-#include <boost/thread/recursive_mutex.hpp>
+#include <boost/thread/mutex.hpp>
 
 namespace {
-  boost::recursive_mutex spinmutex;
+
+/** class to monitor running single-threaded spinners.
+ *
+ *  Calling the callbacks of a callback queue _in order_, requires a unique SingleThreadedSpinner
+ *  spinning on the queue. Other threads accessing the callback queue will probably intercept execution order.
+
+ *  To avoid multiple SingleThreadedSpinners (started from different threads) to operate on the same callback queue,
+ *  this class stores a map of all spinned callback queues.
+ *  If the spinner is single threaded, the corresponding thread-id is stored in the map
+ *  and if other threads will try to spin the same queue, an error message is issued.
+ *
+ *  If the spinner is multi-threaded, the stored thread-id is NULL and future SingleThreadedSpinners
+ *  should not spin this queue.
+ */
+struct SpinnerMonitor
+{
+  /* store spinner information per callback queue:
+     Only as unique single-threaded spinner is allowed and we store its thread id.
+     As multiple multi-threaded spinners are allowed in parallel, their number is counted.
+  */
+  struct Entry
+  {
+    Entry(const boost::thread::id &tid) : tid(tid) {}
+
+    boost::thread::id tid; // thread id of single-threaded spinner
+    unsigned int num_multi_threaded; // number of multi-threaded spinners
```

Comment:
I actually started with this simpler implementation, but turned to the current implementation on purpose:
We want to allow multiple multi-threaded spinners on a queue, but only one single-threaded spinner - in that case also excluding any other multi-threaded spinners.
This will guarantee proper in-order event processing by the single single-threaded spinner.
If there is a multi-threaded spinner acting on a queue, in-order processing obviously isn't considered important by the programmer and thus multiple (multi-threaded) spinners are allowed.